### PR TITLE
Refactor resolver

### PR DESCRIPTION
**Pull request by YuJuncen**
This PR make `resolvers` and `ObserverHandle` together, and make a struct named `TwoPhaseResolver` to prevent resolved ts advancing before initial scanning done.

### CODE REVIEW

File: components/backup-stream/src/endpoint.rs [hunk 1]
```diff
@@ -420,54 +420,42 @@ where
         start_key: Vec<u8>,
         end_key: Vec<u8>,
     ) -> Result<()> {
+        let start_key = start_key;
+        let end_key = end_key;
```
Comment (3pointer) [inline components/backup-stream/src/endpoint.rs]: 🤔

File: components/backup-stream/src/subscription_track.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    }
+}
+
+impl TwoPhaseResolver {
+    pub fn in_phase_one(&self) -> bool {
+        self.stable_ts.is_some()
+    }
+
+    pub fn track_phase_one_lock(&mut self, start_ts: TimeStamp, key: Vec<u8>) {
+        if !self.in_phase_one() {
```
Comment (3pointer) [inline components/backup-stream/src/subscription_track.rs:243]: this seems unreachable?🤔

File: components/backup-stream/src/subscription_track.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+/// - The resolver won't advance the resolved ts to greater than `stable_ts` if there is some. This
+///   can help us prevent resolved ts from advancing when initial scanning hasn't finished yet.
+/// - When we `untrack` a lock haven't been tracked, this would record it, and skip this lock if we want to track it then.
+///   This would be safe because:
+///   - untracking a lock not be tracked is no-op for now.
+///   - tracking a lock have already being untracked (unordered call of `track` and `untrack`) wouldn't happen at phase 2 for same region.
+///     but only when phase 1 and phase 2 happend concurrently, at that time, we wouldn't and cannot advance the resolved ts.
+pub struct TwoPhaseResolver {
+    resolver: Resolver,
+    future_locks: Vec<FutureLock>,
```
Comment (3pointer) [inline components/backup-stream/src/subscription_track.rs:210]: if we use vector here. how to guarantee the `untrack` must appear after `track`.
can we handle the `untrack` happen from raft cmd and the `track` happen from initial scan?🤔

Comment (3pointer) [reply components/backup-stream/src/subscription_track.rs:210]: OK， seems no problem.
we don't put lock in vec from initial sca.

**Review by 3pointer — APPROVED** [review]
LGTM

File: components/backup-stream/src/subscription_track.rs (hunk 2 above)
Comment (YuJuncen) [reply components/backup-stream/src/subscription_track.rs:243]: Just a protection for misuse. 🤔

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**